File: src/main/java/ru/arc/sync/base/SyncRepoBuilder.java
package ru.arc.sync.base;

import ru.arc.network.RedisManager;

import java.util.function.Consumer;
import java.util.function.Function;

public class SyncRepoBuilder<T extends SyncData> {
    Class<T> clazz;
    String key;
    RedisManager redisManager;
    Consumer<T> dataApplier;
    Function<Context, T> dataProducer;

    public SyncRepoBuilder(Class<T> clazz) {
        this.clazz = clazz;
    }

    public SyncRepoBuilder<T> key(String key) {
        this.key = key;
        return this;
    }

    public SyncRepoBuilder<T> redisManager(RedisManager redisManager) {
        this.redisManager = redisManager;
        return this;
    }

    public SyncRepoBuilder<T> dataApplier(Consumer<T> dataSetter) {
        this.dataApplier = dataSetter;
        return this;
    }

    public SyncRepoBuilder<T> dataProducer(Function<Context, T> dataProducer) {
        this.dataProducer = dataProducer;
        return this;
    }


    public SyncRepo<T> build() {
        return new SyncRepo<>(clazz, key, redisManager, dataApplier, dataProducer);
    }
}

File: src/main/kotlin/ru/arc/sync/base/SyncRepoBuilder.kt
package ru.arc.sync.base

import ru.arc.network.RedisManager

class SyncRepoBuilder<T : SyncData>(val clazz: Class<T>) {
    var key: String = ""
    var redisManager: RedisManager? = null
    var dataApplier: ((T) -> Unit)? = null
    var dataProducer: ((Context) -> T?)? = null

    fun key(key: String): SyncRepoBuilder<T> = apply { this.key = key }
    fun redisManager(redisManager: RedisManager): SyncRepoBuilder<T> = apply { this.redisManager = redisManager }
    fun dataApplier(dataSetter: (T) -> Unit): SyncRepoBuilder<T> = apply { this.dataApplier = dataSetter }
    fun dataProducer(dataProducer: (Context) -> T?): SyncRepoBuilder<T> = apply { this.dataProducer = dataProducer }

    fun build(): SyncRepo<T> = SyncRepo(
        clazz = clazz,
        key = key,
        redisManager = checkNotNull(redisManager) { "redisManager must be set" },
        dataApplier = checkNotNull(dataApplier) { "dataApplier must be set" },
        dataProducer = checkNotNull(dataProducer) { "dataProducer must be set" },
    )
}
